velox/exec/GroupingSet.cpp

```diff
@@ -598,6 +606,73 @@ bool GroupingSet::getGlobalAggregationOutput(
   return true;
 }
 
+bool GroupingSet::getDefaultGlobalGroupingSetOutput(
+    int32_t batchSize,
+    bool isPartial,
+    RowContainerIterator& iterator,
+    RowVectorPtr& result) {
+  VELOX_CHECK_LE(globalGroupingSets_->size(), batchSize);
+
+  // Global aggregates don't have grouping keys. But global grouping sets
+  // have null values in grouping keys and a groupId column as well. These
+  // key fields precede the aggregate columns in the result.
+  // This logic builds a row with just aggregate fields to reuse the global
+  // aggregate computation from the regular GroupingSet code-path.
+  auto outputType = asRowType(result->type());
+  auto firstAggregateCol = outputType->children().size() - aggregates_.size();
+  std::vector<std::string> names;
+  std::vector<TypePtr> types;
+  names.reserve(aggregates_.size());
+  types.reserve(aggregates_.size());
+  for (int i = firstAggregateCol; i < outputType->children().size(); i++) {
+    names.push_back(fmt::format("agg{}", i));
+    types.push_back(outputType->childAt(i));
+  }
+  auto aggregatesType =
+      std::make_shared<RowType>(std::move(names), std::move(types));
+  auto globalAggregatesRow = std::dynamic_pointer_cast<RowVector>(
+      BaseVector::create(aggregatesType, 1, &pool_));
+
+  auto status =
+      getGlobalAggregationOutput(1, isPartial, iterator, globalAggregatesRow);
+  if (!status) {
```

Comment:
When would status be false?

Reply:
I'm not entirely sure about 

```
if (iterator.allocationIndex != 0) {
    return false;
  }
```

from getGlobalAggregationOutput function. Can that be removed ?

Reply:
It looks like getGlobalAggregationOutput is designed to be called multiple times. It returns data on first call and 'false' on all subsequent calls. In this use case it should never return false. Let's remove 'status' variable altogether or add VELOX_CHECK(status). If you keep 'status' variable, let's rename for clarity as well.

Reply:
Looking at this some more, I suggest to add the following check at the beginning of this function and assert here that status is true.

```
if (iterator.allocationIndex != 0) {
   return false;
}
```

Reply:
@mbasmanova : I'm might be missing something here. But without this logic to check the repetition the code hangs. getDefaultGlobalAggregationOutput is also called multiple times. The expectation is to return data in the first call and 'false' on subsequent calls. The only way to do this is use the iterator in the same way as getGlobalAggregationOutput.

Any insights ?

Reply:
Fixed as per [other review comment.](https://github.com/facebookincubator/velox/pull/7112#discussion_r1373794393)